### DISCUSSION TITLE
Use V8 FastAPI

### DISCUSSION THREAD

**Discussion by ronag** [uhop/node-re2#218] [Ideas]
https://www.yagiz.co/using-v8-fast-api-in-node-js-core/

V8 Fast API removes a lot of the overhead of calling into C++

**Comment by uhop** [comment]
Thank you for the reference. Obviously I'll study it to understand if it can be leveraged in the project. If this approach is applicable to this extension, it should be applied in a systemic way &mdash; clearly it is not as a singular issue. So I am converting it to a discussion.

**Comment by uhop** [comment]
Is it supported by `NaN`? Node-API? It would be nice to know how to use V8 FastAPI in the context of an addon.

Right now the addon structure is simple: everything is in C++. It practically has no JS. JS started to appear only for advanced features not directly/properly supported by `NaN` like symbols or generators and only as a thin layer. This approach was prompted by the performance concerns.

All methods follow more-or-less the same pattern: decoding parameters from JS to C++ structures, calling the underlying library (RE2), decoding results creating the necessary JS structures. Only the last step involves any JS object creation/memory manipulation.

Currently all of the above are done in C++. As soon as we are in C++ all internal calls are “fast”. My understanding is that we can shift the JS/C++ boundary somehow keeping more logic in JS making “fast” calls to C++ for an advanced functionality. Is this the idea?

I appreciate any input on that topics.

The `replace()` method can accept JS callbacks. Do we have any fast path here when calling JS from C++?

Thanks in advance.

**Comment by uhop** [comment]
It looks like FastAPI is not something we can do now. Postponed. See [V8 FastAPI assessment](https://github.com/uhop/node-re2/wiki/V8-Fast-API-assessment).
